### PR TITLE
fix(sort): reserve Phase-1 input serials under the input lock

### PR DESCRIPTION
**Pull request by nh13**
Fixes a race in the multi-threaded sort engine that can silently truncate sorted output.

## The defect

`try_read_input_blocks` reads a batch of raw BGZF blocks while holding the input-file lock, releases the lock, and only then assigns each block a serial via `input_read_serial.fetch_add` (`crates/fgumi-sort/src/worker_pool.rs:1368-1387`).

A worker preempted between `drop(guard)` and those increments leaves its already-read blocks unaccounted for in `input_read_serial`, while another worker acquires the lock, reads the empty EOF batch, and sets `input_eof`. The done-check — `input_blocks_queued == input_read_serial && input_eof`, documented on `input_blocks_queued` at `worker_pool.rs:693-698` — can then fire early: the merge finalizes and the first worker's trailing blocks are pushed after termination.

The result is a sorted BAM that is **silently short some records, with exit code 0 and no error message**. Because it is timing-dependent it presents as intermittent record loss, and it is on the default multi-threaded path.

## The fix

Reserve the whole batch's serial range with a single `fetch_add` while still holding the lock, so `input_read_serial` accounts for every block that has been read before `input_eof` can be observed as set. Blocks are then pushed under their pre-reserved serials, so ordering is unchanged.

## Testing

There is no deterministic reproducer — the fix rests on the reserve-under-lock invariant, and I would rather say so than imply a regression test pins the race.

The added test is a stress guard: 50k reads spanning many BGZF blocks, sorted at 8 threads with a 1M memory budget to force spills and maximize contention on the input lock, repeated three times. Any lost block surfaces as a record-count mismatch against the input.

- `cargo ci-test`: 5535 passed, 22 skipped
- `cargo ci-fmt`, `cargo ci-lint`: clean

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved multithreaded sorting reliability to avoid premature completion/EOF edge cases that could cause dropped, duplicated, or truncated records.
  * Ensured consistent “record survival” and coordinate-sorted output across large, memory-constrained spill/merge workloads.
* **Tests**
  * Added stronger correctness checks by comparing input/output record identities using order-independent multiset matching.
  * Added a multithreaded stress regression test (currently ignored) to verify no truncation under high contention.

### CODE REVIEW

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

File: crates/fgumi-sort/src/worker_pool.rs [hunk 1]
```diff
@@ -1365,27 +1372,29 @@ impl SortWorkerPool {
...
-        for block in blocks_iter {
-            let serial = shared.input_read_serial.fetch_add(1, Ordering::Relaxed);
-            worker.held_raw_input_blocks.push((serial, block));
-        }
+        dispatch_reserved_blocks(
+            base_serial,
+            blocks,
+            &shared.raw_input_blocks,
+            &mut worker.held_raw_input_blocks,
+        );
```
[the remark this answers is not in the document]
Comment (nh13) [reply crates/fgumi-sort/src/worker_pool.rs:1402]: Fixed, with one deviation from the suggested diff: used `Ordering::AcqRel` rather than `Ordering::Release`.

Every other coordination counter in this struct uses `AcqRel` on its `fetch_add` — `input_blocks_queued` (worker_pool.rs:1317, :1464), `sources_at_eof` (:1752), `decomp_in_flight` (:1719) — so `AcqRel` is the exact sibling pattern, and it also gives the acquire side, which matters here because the returned `base_serial` is only meaningful if this thread also observes prior reservations. That satisfies the point of the finding (no more dependence on an implicit happens-before chain spanning `try_read_input_blocks` and the `input_eof` store) while matching local convention.

While in here I also corrected the justification comment I first wrote alongside the change: it claimed the reservation "must not float past the `input_eof` release store", which is not traceable to this function — both `input_eof.store` sites (:1364 on the I/O-error path, :1371 on the empty-batch EOF path) early-return before reaching the `fetch_add`. The comment now states the real rationale (sibling-counter consistency + explicit visibility at the site).

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review
